### DISCUSSION TITLE
GCC spec files

### DISCUSSION THREAD

**Discussion by i-ky** [clangd/clangd#1042] [Ideas]
I am working on a project that targets GCC and uses [spec files](https://gcc.gnu.org/onlinedocs/gcc/Spec-Files.html). I would like to use `clangd` to index the project. As far as I know, `clang` and related tooling do not support `-specs=...` option, but without it project files can't be parsed correctly.

One possible solution is to manually extract options from spec file and add them to `clangd` configuration file:
```YAML
CompileFlags:
  Add:
 - ...
 - ...
```
Unfortunately, there are `-isystem` options, that are absolute paths depending on the location of project's sources, so these cannot be committed to version control system.

I have considered using [`clang` configuration file](https://clang.llvm.org/docs/UsersManual.html#configuration-files) with `<CFGDIR>`, but it still does not address the issue completely, because there are `-isystem` options which depend on architecture (project's build system generates spec files based on target). Plus there is still some manual work needed to maintain `clang` configuration file in sync.

Since our spec files are used solely for system header options like `-isystem` and `-nostdinc` I had an idea to hack around [`clangd`'s mechanism of system header path detection](https://clangd.llvm.org/guides/system-headers#query-driver). The idea is to whitelist `-specs` and `--specs` [here](https://github.com/llvm/llvm-project/blob/4d006520b8c0cc3a52913b4665bf741c737e5592/clang-tools-extra/clangd/QueryDriverDatabase.cpp#L185-L187)
```C++
  // Preserves these flags and their values, either as separate args or with an
  // equalsbetween them
  const llvm::StringRef ArgsToPreserve[] = {"--sysroot", "-isysroot"};
```
...and add spec file paths to key [here](https://github.com/llvm/llvm-project/blob/4d006520b8c0cc3a52913b4665bf741c737e5592/clang-tools-extra/clangd/QueryDriverDatabase.cpp#L348):
```C++
            QueriedDrivers.get(/*Key=*/(Driver + ":" + Lang).str(), [&] {
```

I have partially implemented the above and a patch seems to solve my problem, now `clangd` can correctly determine system header paths based on `-specs=...` in compilation database. I would like to know your opinion if this feature would be a good addition to `clangd`.

Clearly the lack of spec file support is a problem for GCC-focused projects, but I am not sure if `clangd` in general and `--query-driver` logic in particular is the right place to address this. `-nostdinc` and `-isystem` are not the only options that can be present in spec file. Also it would be great if `-specs` were supported in all `clang`-based tooling, not just `clangd`.

UPD: Related issue #1410.

**Comment by StMartin81** [comment]
This would be great if this would be supported. Just had the same problem and wanted to write something similar before I saw this. :)

**Comment by i-ky** [comment]
https://github.com/llvm/llvm-project/pull/70285 implemented passing `-specs` option(s) to `--query-driver` mechanism.
